zerver/actions/reactions.py

```diff
@@ -82,6 +89,31 @@ def do_add_reaction(
 
     reaction.save()
 
+    # Determine and set the visibility_policy depending on 'automatically_follow_topics_policy'
+    # and 'automatically_unmute_topics_in_muted_streams_policy'.
+    if (
+        message.is_stream_message()
+        and not user_profile.is_bot
+        and user_profile.realm == message.get_realm()
+    ):
+        stream_id = message.recipient.type_id
+        (stream, sub) = access_stream_by_id(user_profile, stream_id)
+        assert stream is not None
+        if sub:
```

Comment:
This might be better with the early return pattern.

Reply:
Or maybe better is to extract this whole block as a function we can call; `user_topics.process_participation(user_profile, message)` as the part inside the `if` statement; and then this could be called both here and in the submessage code path?

Reply:
This works for me!